src/server.rs

```diff
@@ -0,0 +1,137 @@
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use warp::ws::{Message, WebSocket};
+use warp::Filter;
+
+use rtpeeker_common::packet::PacketType;
+use rtpeeker_common::Packet;
+
+use crate::sniffer::Sniffer;
+
+static DIST_DIR: &str = "client/dist";
+static WS_PATH: &str = "ws";
+static NEXT_CLIENT_ID: AtomicUsize = AtomicUsize::new(1);
+
+type Clients = Arc<RwLock<HashMap<usize, mpsc::UnboundedSender<Message>>>>;
+type Packets = Arc<RwLock<Vec<Packet>>>;
+
+pub async fn run<T: pcap::Activated + 'static>(sniffer: Sniffer<T>, addr: SocketAddr) {
+    let packets = Packets::default();
+    let clients = Clients::default();
+
+    let clients_sniff = clients.clone();
+    let packets_sniff = packets.clone();
+
+    tokio::task::spawn(async move {
+        sniff(sniffer, packets_sniff, clients_sniff).await;
+    });
+
+    let packets_filter = warp::any().map(move || packets.clone());
+    let clients_filter = warp::any().map(move || clients.clone());
+    let ws = warp::path(WS_PATH)
+        .and(warp::ws())
+        .and(packets_filter)
+        .and(clients_filter)
+        .map(|ws: warp::ws::Ws, packets, clients| {
+            ws.on_upgrade(move |socket| client_connected(socket, packets, clients))
+        });
+
+    let routes = ws.or(warp::fs::dir(DIST_DIR));
+
+    warp::serve(routes).run(addr).await;
+}
+
+async fn client_connected(ws: WebSocket, packets: Packets, clients: Clients) {
+    let client_id = NEXT_CLIENT_ID.fetch_add(1, Ordering::Relaxed);
+
+    info!("New client connected, assigned id: {}", client_id);
+
+    let (mut client_ws_tx, mut client_ws_rx) = ws.split();
+
+    for pack in packets.read().await.iter() {
+        let Ok(encoded) = pack.encode() else {
+            error!("Failed to encode packet, client_id: {}", client_id);
+            continue;
+        };
+        let msg = Message::binary(encoded);
+        client_ws_tx
+            .feed(msg)
+            .unwrap_or_else(|e| {
+                error!("WebSocket `feed` error: {}, client_id: {}", e, client_id);
+            })
+            .await;
+    }
+
+    client_ws_tx
+        .flush()
+        .unwrap_or_else(|e| {
+            error!("WebSocket `flush` error: {}, client_id: {}", e, client_id);
+        })
+        .await;
+
+    info!(
+        "Sucesfully send already captured packets, client_id: {}",
+        client_id
+    );
+
+    // create channel to send incoming packets to client
+    // and pass it to sniffer via shared state
+    let (tx, rx) = mpsc::unbounded_channel();
+    let mut rx = UnboundedReceiverStream::new(rx);
+
+    tokio::task::spawn(async move {
+        while let Some(message) = rx.next().await {
+            client_ws_tx
+                .send(message)
+                .unwrap_or_else(|e| {
+                    error!("WebSocket `send` error: {}, client_id: {}", e, client_id);
+                })
+                .await;
+        }
+    });
+
+    clients.write().await.insert(client_id, tx);
```

Comment:
I don't get what is going on in this part. Can you explain?
Why can't we just pass sending_channel directly to clients?

Reply:
I guess we could, but that how it's been done in [this example](https://github.com/seanmonstar/warp/blob/master/examples/websockets_chat.rs) this way the unbounded channel is responsible for buffering and flushing the messages (could the WebSocket channel overflow?) and the actual IO is handled by dedicated thread/task, which  may be more efficient.